action.yml

```diff
@@ -207,7 +210,9 @@ runs:
           ${{ inputs.cache-prefix }}${{ runner.os }}-stack-deps-${{ steps.stack-query.outputs.compiler }}-
 
     - name: Dependencies
-      if: steps.restore-deps.outputs.cache-hit != 'true'
+      if: |
+        (success() && steps.restore-deps.outputs.cache-hit != 'true') ||
+        (!cancelled() && inputs.cache-save-always == 'true')
```

Comment:
I worry this doesn't work (the yaml `|` syntax). Isn't this just equivalent to `if: "some long string"`?

I believe the documentation says that `if: ${{ something }}` can be reduced to `if: something` only if `something` is "trivial" or "simple" or something. This seems very not trivial to me, so I'm shocked if it works.

I'll approve, assuming you've checked the example workflows and seen it's working as expected.

Reply:
I think GitHub does some magic to allow this long YAML string to evaluate as a conditional:

When `cache-save-always: false`, a [build failure is observed to bypass saving the cache][0] (see [821cf4d][1], which has the input turned off for `backend-build`). If this was evaluated as a long string, it would always be truthy and so these steps would never no-op. In that same failing build, `cache-save-always` is enabled and the CI run for `backend-test` shows the cache-save step running after the failure. This wouldn't happen for an `if` conditional evaluated as a string since that string would have an implicit `success()` status check function applied.

However, I can't quite verify the default action behavior by removing the build failure and setting `cache-save-always: false`. While the [first CI run for `backend-build`][3] has an expected cache-miss, subsequent builds show the cache key changing, apparently due to the computed `sources-hash` of the cache key. I'm not sure if this has been introduced by this change, but I'll investigate further.

[0]: https://github.com/freckle/megarepo/actions/runs/7618177840/job/20748796709
[1]: https://github.com/freckle/megarepo/pull/32424/commits/821cf4d9c3c4e7c52552cdeb6ccc3dd1290bc0a5
[2]: https://github.com/freckle/megarepo/actions/runs/7618177840/job/20748797130
[3]: https://github.com/freckle/megarepo/actions/runs/7618711349/job/20750418098

Reply:
Made this comment in the wrong thread: https://github.com/freckle/stack-action/pull/34#discussion_r1463861279